lxd/storage/s3/miniod/admin.go

```diff
@@ -0,0 +1,226 @@
+package miniod
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+)
+
+type minioAdmin struct {
+	alias string
+}
+
+// AccountInfo is the response body of the info service account call.
+type AccountInfo struct {
+	ParentUser    string          `json:"parentUser"`
+	AccountStatus string          `json:"accountStatus"`
+	ImpliedPolicy bool            `json:"impliedPolicy"`
+	Policy        json.RawMessage `json:"policy"`
+}
+
+// Credentials holds access and secret keys.
+type Credentials struct {
+	AccessKey    string    `xml:"AccessKeyId"     json:"accessKey"`
+	SecretKey    string    `xml:"SecretAccessKey" json:"secretKey"`
+	SessionToken string    `xml:"SessionToken"    json:"sessionToken"`
+	Expiration   time.Time `xml:"Expiration"      json:"expiration"`
+}
+
+// ServiceAccountArgs is the request options for adding or modifying a service account.
+type ServiceAccountArgs struct {
+	Policy    json.RawMessage `json:"policy"` // Parsed value from iam/policy.Parse()
+	AccessKey string          `json:"accessKey"`
+	SecretKey string          `json:"secretKey"`
+}
+
+// configJSON represents the JSON payload containing the configuration dump.
+type configJSON struct {
+	Status string `json:"status"`
+	Value  []byte `json:"value"`
+}
+
+// NewAdminClient returns a new AdminClient, and assigns the given credentials to an alias, and returns an `mc` client for that alias.
+func NewAdminClient(url string, username string, password string) (*minioAdmin, error) {
+	_, err := shared.RunCommand("mc", "alias", "set", "--insecure", "lxd-minio", api.NewURL().Scheme("http").Host(url).String(), username, password)
```

Comment:
What does this do? Can we instead provide target argument to each invocation rather than requiring client state? Where is that stored?

Reply:
It's the same as setting a LXD remote, in principle. The CLI seems to work exclusively with these aliases as identifiers, but there is a `--config-dir` option that lets us control the location of the state.

Reply:
> Where is that stored?

Reply:
as mc is being called inside the snap

Reply:
It does seem to be smart enough to know to set up a separate state directory inside the snap (the alias isn't registered when running the command from the host), but I think we should be explicit about it anyway.

Reply:
It seems to drop it in `/tmp/.mc` within the snap

Reply:
thats probably ok for the snap, but running lxd outside of the snap now means lxd would modify the existing root user's mc aliases?

Reply:
> It does seem to be smart enough to know to set up a separate state directory inside the snap (the alias isn't registered when running the command from the host), but I think we should be explicit about it anyway.

Yes i think we should be using a path inside lxd's path rather than potentially modifying an external user's aliases (in the case of non-snap installs)

Reply:
I've set the state directory to `/var/snap/lxd/common/lxd/minio`

Reply:
What about non snap installs?

Reply:
I'm using `shared.VarPath` (https://github.com/canonical/lxd/blob/c5337275a370457afdc69480ba453ea89cc152cd/shared/util.go#L219-L230)

So the state dir should be whatever the user is using.

Reply:
OK sounds good thanks